ivadomed/models.py

```diff
@@ -1486,27 +1485,39 @@ def set_model_for_retrain(model_path, retrain_fraction, map_location, reset=True
 def get_model_filenames(folder_model):
     """Get trained model filenames from its folder path.
 
-    This function checks if the folder_model exists and get trained model (.pt or .onnx) and its configuration file
-    (.json) from it.
-    Note: if the model exists as .onnx, then this function returns its onnx path instead of the .pt version.
+    This function checks if the folder_model exists and get trained model path (.pt or .onnx based on
+    model and GPU availability) and its configuration file (.json) from it.
 
     Args:
         folder_name (str): Path of the model folder.
 
     Returns:
-        str, str: Paths of the model (.onnx) and its configuration file (.json).
+        str, str: Paths of the model (.pt or .onnx) and its configuration file (.json).
     """
     if Path(folder_model).is_dir():
+
         prefix_model = Path(folder_model).name
-        # Check if model and model metadata exist. Verify if ONNX model exists, if not try to find .pt model
-        fname_model = Path(folder_model, prefix_model + '.onnx')
-        if not fname_model.is_file():
-            fname_model = Path(folder_model, prefix_model + '.pt')
-            if not fname_model.exists():
-                raise FileNotFoundError(fname_model)
+        fname_model_onnx = Path(folder_model, prefix_model + '.onnx')
+        fname_model_pt = Path(folder_model, prefix_model + '.pt')
+
+        # Assign '.pt' or '.onnx' model based on availability and device.
+        # On GPU, '.pt' model is prioritized whereas the '.onnx' model is prioritized on CPU.
+        if fname_model_onnx.is_file():
+            if fname_model_pt.is_file() and torch.cuda.is_available():
+                fname_model = fname_model_pt
+            else:
+                fname_model = fname_model_onnx
+        elif fname_model_pt.is_file():
+            fname_model = fname_model_pt
+        else:
+            raise FileNotFoundError(f"Model files not found in model folder: "
+                                    f"'{str(fname_model_onnx)}' or '{str(fname_model_pt)}'")
```

Comment:
Hmm... when reading this code, I found that it required a bit of thinking to figure out the different cases. I found myself having to manually trace each section to work out all the possibilities:

1. Neither file exists (error is thrown)
* At least one file exists:
  2. GPU is available, `.pt` exists (_`.pt` used_)
  3. GPU is available, `.pt` does not exist (_`.onnx` used_)
  4. Only CPU is available, `.onnx` exists (_`.onnx` used_)
  5. Only CPU is available, `.onnx` does not exist (_`.pt` used_)

So, one other suggestion I have is to try making these 5 different cases explicit in the code itself:

```python
        # Assign '.pt' or '.onnx' model based on file existence and GPU/CPU device availability
        if not fname_model_pt.is_file() and not fname_model_onnx.is_file():
            raise FileNotFoundError(f"Model files not found in model folder: "
                                    f"'{str(fname_model_onnx)}' or '{str(fname_model_pt)}'")
        # .pt is preferred on GPU, or on CPU if .onnx doesn't exist  
        elif ((    torch.cuda.is_available() and     fname_model_pt.is_file()) or 
              (not torch.cuda.is_available() and not fname_model_onnx.is_file())):
            fname_model = fname_model_pt
        # .onnx is preferred on CPU, or on GPU if .pt doesn't exist 
        elif ((not torch.cuda.is_available() and     fname_model_onnx.is_file()) or 
              (    torch.cuda.is_available() and not fname_model_pt.is_file())):
            fname_model = fname_model_onnx
```

Even though this way contains a lot of redundant checks, my thinking here is that... if the code is self-explanatory, then that will allow a dev to just read the code (without having to spend time inferring the different cases from the code).

Reply:
Thank you again for your excellent suggestion, I agree that it's better to have it self-explanatory. I integrated your suggestion with a small change to call `torch.cuda.is_available()` only once.